openmc/lattice.py

```diff
@@ -759,6 +770,277 @@ def is_valid_index(self, idx):
+                                            attribute)
+                pattern[1, 0] = getattr(self.universes[j][i-1], attribute)
+                if j < self.shape[1] - 1:
+                    pattern[2, 0] = getattr(self.universes[j+1][i-1],
+                                            attribute)
+
+            # Away from bottom edge
+            if j != 0:
+                if i > 0:
+                    pattern[0, 0] = getattr(self.universes[j-1][i-1],
+                                            attribute)
+                pattern[0, 1] = getattr(self.universes[j-1][i], attribute)
+                if i < self.shape[0] - 1:
+                    pattern[0, 2] = getattr(self.universes[j-1][i+1],
+                                            attribute)
+
+            # Away from right edge
+            if i != self.shape[0] - 1:
+                if j > 0:
+                    pattern[0, 2] = getattr(self.universes[j-1][i+1],
+                                            attribute)
+                pattern[1, 2] = getattr(self.universes[j][i+1], attribute)
+                if j < self.shape[1] - 1:
+                    pattern[2, 2] = getattr(self.universes[j+1][i+1],
+                                            attribute)
+
+            # Away from top edge
+            if j != self.shape[1] - 1:
+                if i > 0:
+                    pattern[2, 0] = getattr(self.universes[j+1][i-1],
+                                            attribute)
+                pattern[2, 1] = getattr(self.universes[j+1][i], attribute)
+                if i < self.shape[0] - 1:
+                    pattern[2, 2] = getattr(self.universes[j+1][i+1],
+                                            attribute)
+
+        # Analyze lattice, find unique patterns, ignoring rotations and symmetries
+        for j in range(self.shape[1]):
+            for i in range(self.shape[0]):
+
+                # Skip universes to ignore
+                if self.universes[j][i] in universes_to_ignore:
+                    continue
+
+                # Form the pattern at and around each universe
+                if attribute == "id":
+                    pattern = np.empty(shape=(3, 3), dtype=np.int)
+                if attribute == "name":
+                    pattern = np.empty(shape=(3, 3), dtype=np.dtype('U100'))
+                    if len(self.universes[j][i].name) > 100:
+                        warning.warn("Universe name is larger than 100 "
+                             "characters, which may hinder neighbor search.")
+
+                # Create a neighborhood pattern based on the universe's
+                # neighbors in the grid, and lattice's neighbors at the edges
+
+                # Find neighbors among lattice's neighbors at the edges
+                if len(lattice_neighbors) > 0:
+                    find_edge_neighbors(i, j)
+                else:
+                    pattern[:, :] = dummy_neighbor
```

Comment:
You can move this check for `if len(lattice_neighbors) > 0` into `find_edge_neighbors`. That will make all of the lattice edge logic localized into the `find_edge_neighbors` function.

Reply:
done. You're right it makes more sense